compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -351,6 +359,179 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
+  if (terminator->getNumOperands() != 1) {
+    return false;
+  }
+
+  auto yielded = terminator->getOperand(0).getDefiningOp<arith::NegFOp>();
+  if (!yielded) {
+    return false;
+  }
+
+  if (yielded->getOperand(0) != block->getArgument(0)) {
+    return false;
+  }
+
+  return true;
+}
+
+// This aims to match a partial negation and reverse of a tensor and rewrite
+// it as a single linalg generic. For example, this will match sequences like
+// the following:
+//
+// clang-format off
+//
+//   %extracted_slice = tensor.extract_slice %0[0, ..., 0] [..., 64] [1, ..., 1] : tensor<...x128xf16> to tensor<...x64xf16>
+//   %extracted_slice_0 = tensor.extract_slice %0[0, ..., 64] [..., 64] [1, ..., 1] : tensor<...x128xf16> to tensor<...x64xf16>
+//   %3 = linalg.generic {
+//         indexing_maps = [affine_map<(d0, ..., dn-1) -> (d0, ..., dn-1)>, affine_map<(d0, ..., dn-1) -> (d0, ..., dn-1)>],
+//         iterator_types = [n-1 x "parallel"]}
+//      ins(%extracted_slice_0 : tensor<...x64xf16>) outs(%2 : tensor<...x64xf16>) {
+//   ^bb0(%in: f16, %out: f16):
+//     %5 = arith.negf %in : f16
+//     linalg.yield %5 : f16 
+//   } -> tensor<...x64xf16>
+//   %inserted_slice = tensor.insert_slice %3 into %1[0, ..., 0] [..., 64] [1, ..., 1] : tensor<...x64xf16> into tensor<...x128xf16>
+//   %inserted_slice_1 = tensor.insert_slice %extracted_slice into %inserted_slice[0, ..., 64] [..., 64] [1, ..., 1] : tensor<...x64xf16> into tensor<...x128xf16>
+//
+// clang-format on
+//
+// Where the input tensor is broken down along the inner most dimension, then
+// the bottom elements are negated and the tensor is reconstructed, reversing
+// the order of the slices. This is then rewritten as
+//
+// clang-format off
+//
+//   %expanded = tensor.expand_shape %0 [[0], ..., [n-1, n]] : tensor<...x128xf16> into tensor<...x2x64xf16>
+//   %2 = linalg.generic {
+//         indexing_maps = [affine_map<(d0, ..., dn-1, dn) -> (d0, ..., dn-1, dn)>],
+//         iterator_types = [n x "parallel"]
+//      } outs(%output : tensor<...x2x64xf16>) {
+//   ^bb0(%out: f16):
+//     %i0 = linalg.index 0 : index
+//     ...
+//     %in-1 = linalg.index n-1 : index
+//     %in   = linalg.index n : index
+//     %rev = arith.subi %c1, %in-1 : index
+//     %extracted = tensor.extract %expanded[%i0, ..., %rev, %in] : tensor<...x2x64xf16>
+//     %neg = arith.negf %extracted : f16
+//     %cmp = arith.cmpi eq, %in-1, %c1 : index
+//     %sel = arith.select %cmp, %neg, %extracted : f16
+//     linalg.yield %sel : f16
+//   } -> tensor<...x2x64xf16>
+//   %collapsed = tensor.collapse_shape %2 [[0], ..., [n-1, n]] : tensor<...x2x64xf16> into tensor<...x128xf16>
+//
+// clang-format on
+static std::optional<Value>
+matchCatNegateAndSlice(tensor::InsertSliceOp insertOp) {
+  /// First match against the desired op chain.
+  auto topHalf = insertOp.getSource().getDefiningOp<tensor::ExtractSliceOp>();
+  if (!topHalf || !topHalf->hasOneUse()) {
```

Comment:
I dont think you need to check for one use... even if it doesnt have one use its fine to leave it around for the other use.